CollapsibleGroup with Collapsible children

JarrodMFlesch opened the pull request:
This addresses issue #4 
Allows for natural composition when creating either a `Collapsible` or a `CollapsibleGroup`, and ease of transitioning from independent collapsible's to a group of them.

Also adds tests for both `Collapsible` and `CollapsibleGroup`

JarrodMFlesch committed:
adds children as prop validation, #4

JarrodMFlesch committed:
adds tests for the Collapsible and CollapsibleGroup components

jacobsfletch commented on src/CollapsibleGroup/index.js:
```diff
@@ -55,40 +64,39 @@ class CollapsibleGroup extends Component {
-                )}
-              </Collapsible>
-            );
+          {childrenAsArray.map((collapsible, index) => {
```
`collapsible` as the iteration name is not the most semantic here, since `children` could be something other than an instance of `Collapsible`. It should be renamed to `child`.

i.e. line 99 says `return collapsible`, but is literally anything but a collapsible.

jacobsfletch commented on src/CollapsibleGroup/index.js:
```diff
@@ -55,40 +64,39 @@ class CollapsibleGroup extends Component {
     ].filter(Boolean).join(' ');
 
-    if (collapsibles && collapsibles.length > 0) {
+    const childrenAsArray = React.Children.toArray(children);
```
We should destructure `Children` from React here, as we normally do for `Fragment`, and let's rename `childrenAsArray` to `arrayOfChildren`.

jacobsfletch commented on src/CollapsibleGroup/index.js:
```diff
@@ -108,15 +116,13 @@ CollapsibleGroup.defaultProps = {
   transTime: PropTypes.number,
   transCurve: PropTypes.string,
   classPrefix: PropTypes.string,
+  children: PropTypes.oneOfType([
```
The definition for `children` here can just be `PropTypes.node`, which qualifies both an array of nodes _and_ single nodes.

jacobsfletch commented on demo/Collapsible.demo.js:
```diff
@@ -5,49 +5,41 @@ const CollapsibleDemo = () => (
-        transTime: undefined
-        <br />
-        transCurve: undefined
+        {'<Collapsible'}
```
Let's revert back to straight text here in an effort to make the demo as skeleton as possible. To achieve the same "nesting" affect as the `<ul><li>` structure solves, just use the `&emsp;` HTML entity [as see here](https://github.com/trouble/react-node-position/blob/master/demo/NodeDetails.js) for single or double tabs.

jacobsfletch commented on demo/Collapsible.demo.js:
```diff
@@ -5,49 +5,41 @@ const CollapsibleDemo = () => (
       openOnInit
     >
-      <span>to collapse content</span>
+      <span>beep bop robots</span>
```
Let's use a snippet of lorem ipsum for everything in these demos.

Here's an example of length: `Modnam llam leosed magnis mauris iennam ibulum`

Be sure that each is unique.

jacobsfletch commented on demo/CollapsibleGroup.demo.js:
```diff
@@ -1,67 +1,98 @@
-    />
+    >
+      <Collapsible
+        clickableNode={<span>click me</span>}
```
Bump these props to one line, since they're singular. The ESLint config should probably enforce this.

jacobsfletch commented on src/Collapsible/index.js:
```diff
@@ -72,6 +72,7 @@ class Collapsible extends Component {
               style: {
                 marginBottom: 0,
               },
+              className: `${baseClass}__clickable-node`,
```
I've always disliked the prop name `clickableNode`, even though it's pretty semantic. Any other suggestions?

JarrodMFlesch replied:
@jacobsfletch  what about something with the word `Trigger` in it

jacobsfletch reviewed (changes requested):
Overall, very simple changes (some of which are more-or-less personal preferences).

I do want to be considerate of the commit history, though. I see two commits here with the exact same message, and one insubstantial commit ("small demo text changes") that I'd really like to avoid at all costs. Let's make these revisions in another commit, and then [squash](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) all five down into a single commit.

JarrodMFlesch commented:
IDK how to squash these, I attempted but not sure why it is not working, plz help @jacobsfletch

jacobsfletch committed:
overhauls module composition

- builds CollapsibleToggler component
- builds CollapsibleContent component
- creates CollapsibleContext and useCollapsible hook
- creates CollapsibleGroupContext and useCollapsibleGroup hook
- installs @trbl/react-html-element

jacobsfletch committed:
refines CollapsibleGroup integration

- builds prop hierarchy from Collapsible to CollapsibleGroup, with default fallbacks
- controls all grouped Collapsibles with CollapsibleGroup's toggleCount state

jacobsfletch commented:
@JarrodMFlesch You should check out what I did here. I took what you had started ran with it in a different direction. I think this new composition will be prove to be _quite_ flexible in production.
